datafusion/sql/src/planner.rs

```diff
@@ -336,14 +346,29 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     Some(cte_name.clone()),
                     &mut ctes.clone(),
                     outer_query_schema,
+                    columns,
                 )?;
                 ctes.insert(cte_name, logical_plan);
             }
         }
         let plan = self.set_expr_to_plan(*set_expr, alias, ctes, outer_query_schema)?;
 
         let plan = self.order_by(plan, query.order_by)?;
-        self.limit(plan, query.offset, query.limit)
+        let plan = self.limit(plan, query.offset, query.limit)?;
+        match columns {
+            Some(c) => {
+                let mut new_builder = LogicalPlanBuilder::from(plan.clone());
+                new_builder = new_builder
+                    .project(plan.schema().fields().iter().zip(c.into_iter()).map(
+                        |(field, ident)| col(field.name()).alias(&normalize_ident(ident)),
+                    ))
+                    .map_err(|e| {
+                        context!("Failed to apply alias to inline projection.\n", e)
+                    })?;
+                new_builder.build()
+            }
```

Comment:
We can lose the (mutable) variable here as well:

```suggestion
            Some(c) => LogicalPlanBuilder::from(plan.clone())
                .project(plan.schema().fields().iter().zip(c.into_iter()).map(
                    |(field, ident)| col(field.name()).alias(&normalize_ident(ident)),
                ))
                .map_err(|e| {
                    context!("Failed to apply alias to inline projection.\n", e)
                })?
                .build(),
```